### PR TITLE
Add queue and topic service in service-bus module

### LINKED ISSUE #632: Add Service Bus plan which includes queue/topic
A Service Bus plan should be available in OSBA which includes the creation of a queue or a topic. Ideally such queue (or topic) could be named based on a parameter passed in by the user (not randomly).

### PR DESCRIPTION
**Pull request by norshtein**
Closes #632 
This PR adds `azure-servicebus-queue` and `azure-servicebus-topic` services in service-bus module. Previous service `azure-servicebus` is renamed to `azure-servicebus-namespace`.

### CODE REVIEW

File: pkg/services/servicebus/queue_bind.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+	instance service.Instance,
+	_ service.Binding,
+) (service.Credentials, error) {
+	pdt := instance.Parent.Details.(*namespaceInstanceDetails)
+	dt := instance.Details.(*queueInstanceDetails)
+	return queueCredentials{
+		ConnectionString: string(pdt.ConnectionString),
+		PrimaryKey:       string(pdt.PrimaryKey),
+		QueueURL: fmt.Sprintf(
+			"https://%s.servicebus.windows.net/%s",
```
Comment (zhongyi-zhang) [inline pkg/services/servicebus/queue_bind.go]: Is it useful? I took a look at several SDKs. They need connection string + queue name, e.g. `queueClient = new QueueClient(ServiceBusConnectionString, QueueName);`. And there is no direct queue name in the credentials?

File: docs/modules/servicebus.md [hunk 2]
```diff
@@ -47,7 +48,106 @@ Binding returns the following connection details and shared credentials:
...
+| `maxQueueSize`      | `int`    | The maximum size of the queue in megabytes, which is the size of memory allocated for the queue. | N        | 1024                                                         |
+| `messageTimeToLive` | `string` | ISO 8601 default message timespan to live value. This is the duration after which the message expires, starting from when the message is sent to Service Bus. This is the default value used when TimeToLive is not set on a message itself. For example, `PT276H13M14S` sets the message to expire in 11 day 12 hour 13 minute 14 seconds. | N        | "PT336H"                                                     |
+| `lockDuration`      | `string` | ISO 8601 timespan duration of a peek-lock; that is, the amount of time that the message is locked for other receivers. The lock duration time window can range from 5 seconds to 5 minutes. For example, `PT2M30S` sets the lock duration time to 2 minutes 30 seconds. | N        | "PT30S"                                                      |
+
+##### Bind
+
+Returns a copy of one shared set of credentials.
```
Comment (zhongyi-zhang) [inline docs/modules/servicebus.md]: Is it the right description?

File: docs/modules/servicebus.md [hunk 3]
```diff
@@ -47,7 +48,106 @@ Binding returns the following connection details and shared credentials:
...
+| `parentAlias`       | `string` | Specifies the alias of the namespace in which the  topic should be provisioned. **Note**: the parent must be a service-bus-namespace instance with `standard` or `premium` plan. | Y        |                                                              |
+| `topicName`         | `string` | The name of the topic                                        | N        | If not provided, a random name will be generated as the topic name. |
+| `maxTopicSize`      | `int`    | The maximum size of the topic in megabytes, which is the size of memory allocated for the topic. | N        | 1024                                                         |
+| `messageTimeToLive` | `string` | ISO 8601 default message timespan to live value. This is the duration after which the message expires, starting from when the message is sent to Service Bus. This is the default value used when TimeToLive is not set on a message itself. For example, `PT276H13M14S` sets the message to expire in 11 day 12 hour 13 minute 14 seconds. | N        | "PT336H"                                                     |
+
+##### Bind
+
+Returns a copy of one shared set of credentials.
```
Comment (zhongyi-zhang) [inline docs/modules/servicebus.md]: Some here. And another question: do we create a subscription to the topic for each binding?

File: docs/modules/servicebus.md (hunk 2 above)
Comment (norshtein) [reply docs/modules/servicebus.md]: What's the problem with that...?

**Comment by norshtein** [comment]
Added a binding parameter for topic service, @zhongyi-zhang PTAL.

File: pkg/services/servicebus/queue_bind.go [hunk 4]
```diff
@@ -22,6 +22,7 @@ func (qm *queueManager) GetCredentials(
 	return queueCredentials{
 		ConnectionString: string(pdt.ConnectionString),
 		PrimaryKey:       string(pdt.PrimaryKey),
+		QueueName:        string(dt.ServiceBusQueueName),
 		QueueURL: fmt.Sprintf(
 			"https://%s.servicebus.windows.net/%s",
```
Comment (zhongyi-zhang) [inline pkg/services/servicebus/queue_bind.go]: The sovereign clouds have different suffixes for the endpoint: https://github.com/Azure/go-autorest/blob/master/autorest/azure/environments.go#L102.

Comment (norshtein) [reply pkg/services/servicebus/queue_bind.go]: If so, I would rather remove this field...

File: pkg/services/servicebus/queue_provision.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+func (qm *queueManager) preProvision(
+	ctx context.Context,
+	instance service.Instance,
+) (service.InstanceDetails, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	ppp := instance.Parent.ProvisioningParameters
+	pdt := instance.Parent.Details.(*namespaceInstanceDetails)
+	if queueName := instance.ProvisioningParameters.GetString("queueName"); queueName != "" {
```
[the remark this answers is not in the document]
Comment (norshtein) [reply pkg/services/servicebus/queue_provision.go]: The else scenario is the `return` statement at the end of the function...

File: pkg/services/servicebus/namespace_types.go [hunk 6]
```diff
@@ -2,22 +2,22 @@ package servicebus
 
 import "github.com/Azure/open-service-broker-azure/pkg/service"
 
-type instanceDetails struct {
+type namespaceInstanceDetails struct {
 	ARMDeploymentName       string               `json:"armDeployment"`
 	ServiceBusNamespaceName string               `json:"serviceBusNamespaceName"`
```
[the remark this answers is not in the document]
Comment (norshtein) [reply pkg/services/servicebus/namespace_types.go]: It makes sense. I'll change it.

File: pkg/services/servicebus/queue_bind.go (hunk 4 above)
Comment (zhongyi-zhang) [reply pkg/services/servicebus/queue_bind.go]: I am okay removing it. Or you can modify the connection string to get the correct url. It's up to you.

**Comment by norshtein** [comment]
@zhongyi-zhang Thanks for the reviewing!

**Pull request merged**